src/main/java/org/kairosdb/datastore/cassandra/CassandraDatastore.java

```diff
@@ -656,69 +661,220 @@ private static void filterAndAddKeys(String metricName, SetMultimap<String, Stri
         }
     }
 
+    // TODO remove when old getMatchingRowKeys is uncommented
     private List<DataPointsRowKey> getMatchingRowKeys(String metricName, long startTime, long endTime, SetMultimap<String, String> filterTags, int limit) {
-        final List<DataPointsRowKey> rowKeys = new ArrayList<>();
-        final DataPointsRowKeySerializer keySerializer = new DataPointsRowKeySerializer();
-
+        // determine whether to use split index or not
         String useSplitField = null;
-        Set<String> useSplit = new HashSet<>();
+        Set<String> useSplitSet = new HashSet<>();
 
         for (String split : m_indexTagList) {
             if (filterTags.containsKey(split)) {
                 Set<String> vs = filterTags.get(split);
-                if (((vs.size() < useSplit.size() && useSplit.size() > 0) || (vs.size() > 0 && useSplit.size() == 0))
+                if (((vs.size() < useSplitSet.size() && useSplitSet.size() > 0) || (vs.size() > 0 && useSplitSet.size() == 0))
                         && vs.stream().noneMatch(x -> x.contains("*") || x.contains("?"))) {
-                    useSplit = vs;
+                    useSplitSet = vs;
                     useSplitField = split;
                 }
             }
         }
+        List<String> useSplit = new ArrayList<>(useSplitSet);
```

Comment:
Is it a restriction of BoundStatement below that requires us to do this?  Underneath this will be an array copy of everything, so it could get expensive.

Reply:
and that's exactly how it was before, so at least it doesn't make things worse :). But you're right, `BoundStatement` has  `setSet` method, so we can try to avoid conversion